src/pages/signin/SignInPage.js

```diff
@@ -51,37 +57,65 @@ class SignInPage extends Component {
         // - AND is not validated or password is forgotten
-        const shouldShowResendValidationLinkForm = this.props.credentials.login
-            && (!this.props.account.validated || this.props.account.forgotPassword);
+        // - AND user is not on 'passwordless' beta
+        const showSendValidateCodeForm = this.props.credentials.login
+            && (!this.props.account.validated || this.props.account.forgotPassword)
+            && !Permissions.canUsePasswordlessLogins(this.props.betas);
 
-        const welcomeText = shouldShowResendValidationLinkForm
-            ? ''
-            : this.props.translate(`welcomeText.${showPasswordForm ? 'welcomeBack' : 'welcome'}`);
+        let welcomeText;
+        if (showValidateCodeForm) {
+            if (this.props.account.requiresTwoFactorAuth) {
+                // We will only know this after a user signs in successfully, without their 2FA code
+                welcomeText = this.props.translate('validateCodeForm.enterTwoFactorOrRecoveryCode');
+            } else if (this.props.account.validated) {
+                welcomeText = this.props.translate('welcomeText.welcomeBackEnterMagicCode', {login: this.props.credentials.login});
+            } else if (!this.props.account.validated) {
+                welcomeText = this.props.translate('welcomeText.welcomeEnterMagicCode', {login: this.props.credentials.login});
+            }
+        } else if (showSendValidateCodeForm) {
+            welcomeText = '';
+        } else if (showPasswordForm) {
+            welcomeText = this.props.translate('welcomeText.welcomeBack');
+        } else {
+            welcomeText = this.props.translate('welcomeText.welcome');
+        }
 
         return (
             <SafeAreaView style={[styles.signInPage]}>
                 <SignInPageLayout
                     welcomeText={welcomeText}
-                    shouldShowWelcomeText={showLoginForm || showPasswordForm || !shouldShowResendValidationLinkForm}
+                    shouldShowWelcomeText={showLoginForm || showPasswordForm || showValidateCodeForm || !showSendValidateCodeForm}
                 >
                     {/* LoginForm and PasswordForm must use the isVisible prop. This keeps them mounted, but visually hidden
                     so that password managers can access the values. Conditionally rendering these components will break this feature. */}
                     <LoginForm isVisible={showLoginForm} blurOnSubmit={this.props.account.validated === false} />
-                    <PasswordForm isVisible={showPasswordForm} />
-                    {shouldShowResendValidationLinkForm && <ResendValidationForm />}
+                    {showValidateCodeForm ? (
+                        <ValidateCodeForm isVisible={showValidateCodeForm} />
+                    ) : (
+                        <PasswordForm isVisible={showPasswordForm} />
+                    )}
+                    {showSendValidateCodeForm && <ResendValidationForm />}
```

Comment:
NAB: might be nice to rename `showSendValidateCodeForm` to `showResendValidationForm` since it's so close to `showValidateCodeForm` and the form name is not `SendValidateCodeForm`

Reply:
Definitely like this 👍